# Household Energy Forecasting

## 1. Importing Libraries

In [40]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option("display.max_columns", None)

## 2. Loading Dataset

In [41]:
file_path = "household_power_consumption.txt"

df = pd.read_csv(
    file_path,
    sep=";",
    na_values="?",
    low_memory=False
)

df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,16/12/2006,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0
1,16/12/2006,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0
2,16/12/2006,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0
3,16/12/2006,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0
4,16/12/2006,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0


In [42]:
print("Dataset shape:", df.shape)
print("\nColumn names:")
print(df.columns.tolist())

df.info()

Dataset shape: (2075259, 9)

Column names:
['Date', 'Time', 'Global_active_power', 'Global_reactive_power', 'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2', 'Sub_metering_3']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Date                   object 
 1   Time                   object 
 2   Global_active_power    float64
 3   Global_reactive_power  float64
 4   Voltage                float64
 5   Global_intensity       float64
 6   Sub_metering_1         float64
 7   Sub_metering_2         float64
 8   Sub_metering_3         float64
dtypes: float64(7), object(2)
memory usage: 142.5+ MB


## 3. Datetime Preparation

In [43]:
# Combine Date and Time into one datetime column
df["Datetime"] = pd.to_datetime(
    df["Date"] + " " + df["Time"],
    format="%d/%m/%Y %H:%M:%S"
)

df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,Datetime
0,16/12/2006,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0,2006-12-16 17:24:00
1,16/12/2006,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0,2006-12-16 17:25:00
2,16/12/2006,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0,2006-12-16 17:26:00
3,16/12/2006,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0,2006-12-16 17:27:00
4,16/12/2006,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0,2006-12-16 17:28:00


In [44]:
# Set Datetime as the index
df = df.set_index("Datetime")

df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
Datetime,,,,,,,,,
2006-12-16 17:24:00,16/12/2006,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0
2006-12-16 17:25:00,16/12/2006,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0
2006-12-16 17:26:00,16/12/2006,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0
2006-12-16 17:27:00,16/12/2006,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0
2006-12-16 17:28:00,16/12/2006,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0


In [45]:
# Remove the original Date and Time columns
df = df.drop(columns=["Date", "Time"])

df.head()

,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
Datetime,,,,,,,
2006-12-16 17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0
2006-12-16 17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0
2006-12-16 17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0
2006-12-16 17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0
2006-12-16 17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0


In [46]:
print("Index type:", type(df.index))
print("Start date:", df.index.min())
print("End date:", df.index.max())
print("Chronologically sorted:", df.index.is_monotonic_increasing)
print("Duplicate timestamps:", df.index.duplicated().sum())

df.info()

Index type: <class 'pandas.core.indexes.datetimes.DatetimeIndex'>
Start date: 2006-12-16 17:24:00
End date: 2010-11-26 21:02:00
Chronologically sorted: True
Duplicate timestamps: 0
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2075259 entries, 2006-12-16 17:24:00 to 2010-11-26 21:02:00
Data columns (total 7 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Global_active_power    float64
 1   Global_reactive_power  float64
 2   Voltage                float64
 3   Global_intensity       float64
 4   Sub_metering_1         float64
 5   Sub_metering_2         float64
 6   Sub_metering_3         float64
dtypes: float64(7)
memory usage: 126.7 MB


### Why DatetimeIndex?

Date and time were combined into a DatetimeIndex so the readings could be ordered, resampled by hour, split chronologically, and analyzed using daily and weekly time patterns.

## 4. Missing-Value Analysis

In [47]:
# Count missing values in each column
missing_count = df.isna().sum()

# Calculate the percentage missing in each column
missing_percent = (missing_count / len(df)) * 100

missing_summary = pd.DataFrame({
    "Missing Values": missing_count,
    "Missing Percentage": missing_percent
})

missing_summary

,Missing Values,Missing Percentage
Global_active_power,25979,1.251844
Global_reactive_power,25979,1.251844
Voltage,25979,1.251844
Global_intensity,25979,1.251844
Sub_metering_1,25979,1.251844
Sub_metering_2,25979,1.251844
Sub_metering_3,25979,1.251844


In [48]:
# Count rows containing at least one missing value
rows_with_missing = df.isna().any(axis=1).sum()

print("Rows with at least one missing value:", rows_with_missing)
print(
    "Percentage of rows with missing values:",
    round(rows_with_missing / len(df) * 100, 4),
    "%"
)

Rows with at least one missing value: 25979
Percentage of rows with missing values: 1.2518 %


In [49]:
missing_per_row = df.isna().sum(axis=1)

print("Missing-column count per affected row:")
print(missing_per_row[missing_per_row > 0].value_counts().sort_index())

Missing-column count per affected row:
7    25979
Name: count, dtype: int64


### 4.1 Minute-Level Gap Analysis

Consecutive missing timestamps are grouped to measure the number and duration of sensor outages. This helps determine whether one imputation method is suitable for all gaps.

In [50]:
missing_rows = df.isna().any(axis=1)

gap_groups = missing_rows.ne(missing_rows.shift()).cumsum()

missing_gap_lengths = (
    missing_rows[missing_rows]
    .groupby(gap_groups[missing_rows])
    .size()
)

print("Number of separate missing gaps:", len(missing_gap_lengths))
print("Shortest gap:", missing_gap_lengths.min(), "minutes")
print("Longest gap:", missing_gap_lengths.max(), "minutes")
print("Median gap:", missing_gap_lengths.median(), "minutes")
print("Mean gap:", round(missing_gap_lengths.mean(), 2), "minutes")

print("\nMissing-gap length distribution:")
print(missing_gap_lengths.describe())

Number of separate missing gaps: 71
Shortest gap: 1 minutes
Longest gap: 7226 minutes
Median gap: 1.0 minutes
Mean gap: 365.9 minutes

Missing-gap length distribution:
count      71.000000
mean      365.901408
std      1251.468043
min         1.000000
25%         1.000000
50%         1.000000
75%         3.000000
max      7226.000000
dtype: float64


In [51]:
missing_gap_lengths.sort_values(ascending=False).head(10)

138    7226
140    5237
14     3723
100    3305
118    3129
126    2027
104     891
32       83
78       70
34       47
dtype: int64

## 5. Hourly Resampling

In [52]:
hourly_df = df.resample("h").mean()

valid_readings_per_hour = df["Global_active_power"].resample("h").count()

hourly_df.loc[valid_readings_per_hour < 30] = np.nan
# At least 30 valid minute readings are required for an hourly value.
print("Minute-level shape:", df.shape)
print("Hourly shape:", hourly_df.shape)
print("Start:", hourly_df.index.min())
print("End:", hourly_df.index.max())

hourly_df.head()

Minute-level shape: (2075259, 7)
Hourly shape: (34589, 7)
Start: 2006-12-16 17:00:00
End: 2010-11-26 21:00:00


,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
Datetime,,,,,,,
2006-12-16 17:00:00,4.222889,0.229000,234.643889,18.100000,0.0,0.527778,16.861111
2006-12-16 18:00:00,3.632200,0.080033,234.580167,15.600000,0.0,6.716667,16.866667
2006-12-16 19:00:00,3.400233,0.085233,233.232500,14.503333,0.0,1.433333,16.683333
2006-12-16 20:00:00,3.268567,0.075100,234.071500,13.916667,0.0,0.000000,16.783333
2006-12-16 21:00:00,3.056467,0.076667,237.158667,13.046667,0.0,0.416667,17.216667


In [53]:
hourly_missing_summary = pd.DataFrame({
    "Missing Hours": hourly_df.isna().sum(),
    "Missing Percentage": hourly_df.isna().mean() * 100
})

hourly_missing_summary

,Missing Hours,Missing Percentage
Global_active_power,432,1.248952
Global_reactive_power,432,1.248952
Voltage,432,1.248952
Global_intensity,432,1.248952
Sub_metering_1,432,1.248952
Sub_metering_2,432,1.248952
Sub_metering_3,432,1.248952


In [54]:
print(
    "Hours with at least one missing value:",
    hourly_df.isna().any(axis=1).sum()
)

print(
    "Total hourly rows:",
    len(hourly_df)
)

Hours with at least one missing value: 432
Total hourly rows: 34589


### 5.1 Hourly Gap Analysis

The remaining missing hourly observations are grouped into consecutive gaps to determine which outages are short enough for interpolation and which require a seasonal replacement method.

In [55]:
missing_hour_rows = hourly_df.isna().any(axis=1)

# Group consecutive missing and non-missing hours
hour_gap_groups = missing_hour_rows.ne(
    missing_hour_rows.shift()
).cumsum()

hourly_gap_lengths = (
    missing_hour_rows[missing_hour_rows]
    .groupby(hour_gap_groups[missing_hour_rows])
    .size()
)

print("Number of separate hourly gaps:", len(hourly_gap_lengths))
print("Shortest gap:", hourly_gap_lengths.min(), "hours")
print("Longest gap:", hourly_gap_lengths.max(), "hours")
print("Median gap:", hourly_gap_lengths.median(), "hours")
print("Mean gap:", round(hourly_gap_lengths.mean(), 2), "hours")

print("\nHourly gap summary:")
print(hourly_gap_lengths.describe())

Number of separate hourly gaps: 13
Shortest gap: 1 hours
Longest gap: 120 hours
Median gap: 15.0 hours
Mean gap: 33.23 hours

Hourly gap summary:
count     13.000000
mean      33.230769
std       39.344957
min        1.000000
25%        1.000000
50%       15.000000
75%       55.000000
max      120.000000
dtype: float64


### 5.2 Resampling and Imputation Decision

After resampling, 432 of 34,589 hourly observations remained missing across 13 gaps ranging from 1 to 120 hours.

So we selected a hybrid strategy to handle this

- Gaps of up to 3 hours are filled using time interpolation.
- Longer gaps are filled using the training-set median for the same hour of the week.


In [56]:
print(
    "Valid readings in final hour:",
    valid_readings_per_hour.loc["2010-11-26 21:00:00"]
)

print("\nFinal hourly row:")
print(hourly_df.loc["2010-11-26 21:00:00"])

Valid readings in final hour: 3

Final hourly row:
Global_active_power     NaN
Global_reactive_power   NaN
Voltage                 NaN
Global_intensity        NaN
Sub_metering_1          NaN
Sub_metering_2          NaN
Sub_metering_3          NaN
Name: 2010-11-26 21:00:00, dtype: float64


## 6. Chronological Train, Validation, and Test Split

In [57]:
# Define split boundaries
test_start = hourly_df.index.max() - pd.DateOffset(months=6)
val_start = test_start - pd.DateOffset(months=3)

train_df = hourly_df.loc[hourly_df.index < val_start].copy()
val_df = hourly_df.loc[
    (hourly_df.index >= val_start) &
    (hourly_df.index < test_start)
].copy()
test_df = hourly_df.loc[hourly_df.index >= test_start].copy()

print("Training:", train_df.index.min(), "to", train_df.index.max())
print("Validation:", val_df.index.min(), "to", val_df.index.max())
print("Testing:", test_df.index.min(), "to", test_df.index.max())

print("\nShapes:")
print("Train:", train_df.shape)
print("Validation:", val_df.shape)
print("Test:", test_df.shape)

print("\nMissing rows:")
print("Train:", train_df.isna().any(axis=1).sum())
print("Validation:", val_df.isna().any(axis=1).sum())
print("Test:", test_df.isna().any(axis=1).sum())

Training: 2006-12-16 17:00:00 to 2010-02-26 20:00:00
Validation: 2010-02-26 21:00:00 to 2010-05-26 20:00:00
Testing: 2010-05-26 21:00:00 to 2010-11-26 21:00:00

Shapes:
Train: (28036, 7)
Validation: (2136, 7)
Test: (4417, 7)

Missing rows:
Train: 190
Validation: 34
Test: 208


In [58]:
# Create hour of week feature
def add_hour_of_week(frame):
    result = frame.copy()

    result["hour_of_week"] = (
        result.index.dayofweek * 24
        + result.index.hour
    )

    return result

In [59]:
# Add hour of week feature to each split
train_temp = add_hour_of_week(train_df)
val_temp = add_hour_of_week(val_df)
test_temp = add_hour_of_week(test_df)

In [60]:
# Store the feature columns
feature_columns = train_df.columns.tolist()

feature_columns

['Global_active_power',
 'Global_reactive_power',
 'Voltage',
 'Global_intensity',
 'Sub_metering_1',
 'Sub_metering_2',
 'Sub_metering_3']

In [61]:
# Calculate the median of each feature in training
training_medians = train_df.median()

training_medians

Global_active_power        0.817250
Global_reactive_power      0.104567
Voltage                  240.901333
Global_intensity           3.488333
Sub_metering_1             0.000000
Sub_metering_2             0.316667
Sub_metering_3             1.616667
dtype: float64

In [62]:
# Calculate the median of each feature for each hour of the week in training data
hour_of_week_medians = (
    train_temp
    .groupby("hour_of_week")[feature_columns]
    .median()
)

hour_of_week_medians.head()

,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
hour_of_week,,,,,,,
0,0.364033,0.102033,242.917833,1.586667,0.0,0.333333,0.666667
1,0.330200,0.091833,242.331833,1.426667,0.0,0.216667,0.666667
2,0.316733,0.096133,243.114833,1.406667,0.0,0.333333,0.650000
3,0.315067,0.090100,243.044833,1.403333,0.0,0.216667,0.650000
4,0.315000,0.087300,242.412000,1.400000,0.0,0.216667,0.666667


## 7. Missing-Value Imputation

In [63]:
# Impute missing values in each split using median values for each hour of the week
def impute_hourly_data(frame):
    filled = frame.copy()
# Fill missing values using the median for each hour of the week

    filled = filled.interpolate(
        method="time",
        limit=3,
        limit_area="inside"
    )
# Add hour of week feature to the filled DataFrame
    hour_of_week = (
        filled.index.dayofweek * 24
        + filled.index.hour
    )
# Fill remaining long gaps using the median for each hour of the week
    for column in feature_columns:
        missing_mask = filled[column].isna()

        filled.loc[missing_mask, column] = (
            hour_of_week_medians.loc[
                hour_of_week[missing_mask],
                column
            ].to_numpy()
        )
# Add Fallback and return Result
    filled = filled.fillna(training_medians)

    return filled

In [64]:
# Impute missing values in each split
train_imputed = impute_hourly_data(train_df)
val_imputed = impute_hourly_data(val_df)
test_imputed = impute_hourly_data(test_df)

print("Missing values after imputation:")
print("Train:", train_imputed.isna().sum().sum())
print("Validation:", val_imputed.isna().sum().sum())
print("Test:", test_imputed.isna().sum().sum())

Missing values after imputation:
Train: 0
Validation: 0
Test: 0


In [65]:
print("Shapes:")
print("Train:", train_imputed.shape)
print("Validation:", val_imputed.shape)
print("Test:", test_imputed.shape)

print("\nDate ranges:")
print("Train:", train_imputed.index.min(), "to", train_imputed.index.max())
print("Validation:", val_imputed.index.min(), "to", val_imputed.index.max())
print("Test:", test_imputed.index.min(), "to", test_imputed.index.max())

Shapes:
Train: (28036, 7)
Validation: (2136, 7)
Test: (4417, 7)

Date ranges:
Train: 2006-12-16 17:00:00 to 2010-02-26 20:00:00
Validation: 2010-02-26 21:00:00 to 2010-05-26 20:00:00
Test: 2010-05-26 21:00:00 to 2010-11-26 21:00:00


## 8. Feature Scaling

MinMax normalization is applied because LSTM models train more effectively when input features are on comparable numeric scales.

The scaler is fitted only on the training data. The same fitted scaler is then used to transform the validation and test sets

In [66]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

train_scaled = scaler.fit_transform(train_imputed)
val_scaled = scaler.transform(val_imputed)
test_scaled = scaler.transform(test_imputed)

print("Train scaled shape:", train_scaled.shape)
print("Validation scaled shape:", val_scaled.shape)
print("Test scaled shape:", test_scaled.shape)

Train scaled shape: (28036, 7)
Validation scaled shape: (2136, 7)
Test scaled shape: (4417, 7)


## 9. LSTM Sequence Preparation

Hourly data is converted into supervised learning sequences. Each input sample contains 168 hourly time steps, representing the previous 7 days, across all 7 sensor features.

In [67]:
target_column = "Global_active_power"
target_index = feature_columns.index(target_column)

print("Target index:", target_index)

Target index: 0


### 9.1 Sliding-Window Function

In [68]:
def create_sequences(
    data,
    input_steps=168,
    output_steps=24,
    target_index=0
):
    X, y = [], []

    for i in range(len(data) - input_steps - output_steps + 1):
        X.append(data[i:i + input_steps])

        y.append(
            data[
                i + input_steps:
                i + input_steps + output_steps,
                target_index
            ]
        )

    return np.array(X), np.array(y)

In [69]:
X_train, y_train = create_sequences(
    train_scaled,
    target_index=target_index
)

X_val, y_val = create_sequences(
    val_scaled,
    target_index=target_index
)

X_test, y_test = create_sequences(
    test_scaled,
    target_index=target_index
)

In [70]:
print("X_train:", X_train.shape)
print("y_train:", y_train.shape)

print("X_val:", X_val.shape)
print("y_val:", y_val.shape)

print("X_test:", X_test.shape)
print("y_test:", y_test.shape)

X_train: (27845, 168, 7)
y_train: (27845, 24)
X_val: (1945, 168, 7)
y_val: (1945, 24)
X_test: (4226, 168, 7)
y_test: (4226, 24)


The resulting input arrays have the shape:

`(samples, 168 time steps, 7 features)`

The target arrays have the shape:

`(samples, 24 forecast hours)`

## 10. LSTM Model Development

In [71]:
import tensorflow as tf

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, LSTM, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping

In [72]:
# Set random seeds for more reproducible results
np.random.seed(42)
tf.random.set_seed(42)
# Build the stacked LSTM architecture
model = Sequential([
    Input(shape=(168, 7)),  # Seven days of hourly history across seven features
    LSTM(64, return_sequences=True),    # Return the entire temporal sequence for the next LSTM layer
    Dropout(0.2),

    # Convert the sequence into a compact temporal representation
    LSTM(32),
    Dropout(0.2),

    # Predict the next 24 hourly target values
    Dense(24)
])

In [73]:
# Compile the model
model.compile(
    optimizer="adam",
    loss="mse",
    metrics=["mae"]
)

model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ lstm_2 (LSTM)                   │ (None, 168, 64)        │        18,432 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 168, 64)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_3 (LSTM)                   │ (None, 32)             │        12,416 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 32)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 24)             │           792 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 31,640 (123.59 KB)

 Trainable params: 31,640 (123.59 KB)

 Non-trainable params: 0 (0.00 B)

## 11. Model Training

The model is trained on the chronological training sequences and evaluated on the validation set after each epoch.

In [74]:
# Stop training when validation loss does not improve for 5 epochs
early_stopping = EarlyStopping(
    monitor="val_loss",
    patience=5,
    restore_best_weights=True
)

In [ ]:
# Train the LSTM model
history = model.fit(
    X_train,
    y_train,
    validation_data=(X_val, y_val),
    epochs=50,
    batch_size=32,
    callbacks=[early_stopping],

    # Preserve chronological order
    shuffle=False
)

### 11.1 Training History

In [ ]:
plt.figure(figsize=(10, 5))

plt.plot(history.history["loss"], label="Training Loss")
plt.plot(history.history["val_loss"], label="Validation Loss")

plt.xlabel("Epoch")
plt.ylabel("Mean Squared Error")
plt.title("LSTM Training and Validation Loss")
plt.legend()
plt.grid(True)
plt.show()

In [ ]:
plt.figure(figsize=(10, 5))

plt.plot(history.history["mae"], label="Training MAE")
plt.plot(history.history["val_mae"], label="Validation MAE")

plt.xlabel("Epoch")
plt.ylabel("Mean Absolute Error")
plt.title("LSTM Training and Validation MAE")
plt.legend()
plt.grid(True)
plt.show()